mithril-common/src/cardano_transaction_parser.rs

```diff
@@ -119,48 +115,60 @@ impl CardanoTransactionParser {
 
     /// Read blocks from immutable file
     fn read_blocks_from_immutable_file(immutable_file: &ImmutableFile) -> StdResult<Vec<Block>> {
-        let cbor = fs::read(&immutable_file.path).with_context(|| {
-            format!(
-                "CardanoTransactionParser could not read from immutable file: {:?}",
-                immutable_file.path
-            )
-        })?;
-
-        let mut blocks_start_byte_index: Vec<_> = (0..cbor.len())
-            .filter(|&byte_index| {
-                let cbor_header_maybe = &cbor[byte_index..min(byte_index + 2, cbor.len())];
-                match block_era(cbor_header_maybe) {
-                    Outcome::Matched(_) | Outcome::EpochBoundary => true,
-                    Outcome::Inconclusive => false,
-                }
-            })
-            .collect();
-
-        blocks_start_byte_index.push(cbor.len() + 1);
+        let hardano_blocks = hardano_blocks(immutable_file)?;
 
         let mut blocks = Vec::new();
-        let mut last_start_byte_index = 0;
-        for block_start_index in blocks_start_byte_index.into_iter().skip(1) {
-            let maybe_end_byte_index = min(block_start_index, cbor.len());
-            if let Ok(multi_era_block) =
-                MultiEraBlock::decode(&cbor[last_start_byte_index..maybe_end_byte_index])
-            {
-                let block = Block::try_convert(multi_era_block, immutable_file.number)
-                    .with_context(|| {
-                        format!(
-                        "CardanoTransactionParser could not read data from block in immutable file: {:?}",
+        for block in hardano_blocks {
+            let block = block.map_err(|e| {
+                anyhow!(e).context(format!(
+                    "Error while reading block in immutable file: '{:?}'",
+                    immutable_file.path
+                ))
+            })?;
+
+            match MultiEraBlock::decode(&block) {
+                Ok(multi_era_block) => {
+                    let block = Block::try_convert(multi_era_block, immutable_file.number)
+                        .with_context(|| {
+                            format!(
+                            "CardanoTransactionParser could not read data from block in immutable file: {:?}",
+                            immutable_file.path
+                        )
+                        })?;
+                    blocks.push(block);
+                }
+                Err(err) => {
+                    return Err(anyhow!(err).context(format!(
+                        "Error while decoding block in immutable file: '{:?}'",
                         immutable_file.path
-                    )
-                    })?;
-                blocks.push(block);
-                last_start_byte_index = block_start_index;
+                    )))
+                }
             }
```

Comment:
This could probably be more readable by using a `map_err` on the `MultiEraBlock::decode(&block)` as in `let multi_era_block = MultiEraBlock::decode(&block).map_er(...)?;`.

Reply:
We also extract a function to separate iteration from block conversion.